File: 94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end

# Recursive solution
def inorder_traversal(root)
    result = []
    inorder_traversal_helper(root, result)
    result
end

def inorder_traversal_helper(node, result)
    return if node.nil?
    
    inorder_traversal_helper(node.left, result)
    result << node.val
    inorder_traversal_helper(node.right, result)
end

# Iterative solution
def inorder_traversal_iterative(root)
    result = []
    stack = []
    current = root
    
    while !current.nil? || !stack.empty?
        while !current.nil?
            stack.push(current)
            current = current.left
        end
        
        current = stack.pop
        result << current.val
        current = current.right
    end
    
    result
end

File: 94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  // Recursive solution
  List<int> inorderTraversal(TreeNode? root) {
    List<int> result = [];
    _inorderTraversalHelper(root, result);
    return result;
  }
  
  void _inorderTraversalHelper(TreeNode? node, List<int> result) {
    if (node == null) return;
    
    _inorderTraversalHelper(node.left, result);
    result.add(node.val);
    _inorderTraversalHelper(node.right, result);
  }
  
  // Iterative solution
  List<int> inorderTraversalIterative(TreeNode? root) {
    List<int> result = [];
    List<TreeNode> stack = [];
    TreeNode? current = root;
    
    while (current != null || stack.isNotEmpty) {
      while (current != null) {
        stack.add(current);
        current = current.left;
      }
      
      current = stack.removeLast();
      result.add(current.val);
      current = current.right;
    }
    
    return result;
  }
}